### ISSUE TITLE
HWISignature should contain a Vec<u8> instead of String

### ISSUE THREAD

**Issue by afilini** [bitcoindevkit/rust-hwi#4]
As for #3, it could implement `Deref<Target = Vec<u8>>` or even `AsRef<[u8]>`

**Comment by tanishq67** [comment]
Hi, thank you for raising up the issue.
Do you think this is fine?
![image](https://user-images.githubusercontent.com/52757132/154262537-47babc08-4776-41ee-a4f8-03c41bd8b730.png)

**Comment by afilini** [comment]
Hi, not really, no, the issue title mentioned to switch to `Vec<u8>`, not `Vec<u32>`.

But actually there's more to implement, because just changing the type is not enough. HWI returns a string for a signature which I believe contains the hex value, so you would also have to annotate the structure so that serde knows how to turn that hex string into a `Vec<u8>` (which is not that hard, but you've gotta be a bit familiar with rust before you can do that).

Also, a suggestion: next time just copy-paste the code instead of taking a screenshot, it makes it easier for other people to copy-paste it again if the need to try something out rather than having to type it manually. See the [markdown docs](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code) for more.

**Comment by tanishq67** [comment]
Thank You very much for your reply. 
> Hi, not really, no, the issue title mentioned to switch to `Vec<u8>`, not `Vec<u32>`.
I switched to u32 rather than u8 because I saw that every vector was in form of 32 bit number so keep the code consistent I did that.

> But actually there's more to implement, because just changing the type is not enough. HWI returns a string for a signature which I believe contains the hex value, so you would also have to annotate the structure so that serde knows how to turn that hex string into a `Vec<u8>` (which is not that hard, but you've gotta be a bit familiar with rust before you can do that).

Thanks again for the insights I will try to work upon it.

> Also, a suggestion: next time just copy-paste the code instead of taking a screenshot, it makes it easier for other people to copy-paste it again if the need to try something out rather than having to type it manually. See the [markdown docs](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code) for more.
Yeah, sorry my bad.

**Comment by tanishq67** [comment]
Hello, sorry for the delay, I had my midterm university exam and had to get myself familiar with Rust.
I came up with this solution. It turns the hex string into Vec<u8>. Let me know your thoughts upon it.
```
use serde::{Deserialize};

#[derive(Deserialize, Debug)]
pub struct HWISignature {
    #[serde(with="hex_to_u8")]
    pub signature: Vec<u8>
}

mod hex_to_u8 {
    use serde::{Deserialize};
    use serde::{Deserializer};
    use std::num::ParseIntError;

    pub fn decode_hex(s: &str) -> Result<Vec<u8>, ParseIntError> {
        (0..s.len())
            .step_by(2)
            .map(|i| u8::from_str_radix(&s[i..i + 2], 16))
            .collect()
    }

    pub fn deserialize<'de, D: Deserializer<'de>>(d: D) -> Result<Vec<u8>, D::Error> {
        let hex_to_u8 = String::deserialize(d)?;
        let hex_string = std::str::from_utf8(hex_to_u8.as_bytes()).unwrap().to_string();
        decode_hex(&hex_string)
        .map_err(|e| serde::de::Error::custom(e))
    }
}

fn main() {
    let j = "
        {
            \"signature\": \"464F4F424152\"
        }";
        // FOOBAR
    let u: HWISignature = serde_json::from_str(j).unwrap();
    println!("{:#?}", u);
}
```
**Output**
```
HWISignature {
    signature: [
        70,
        79,
        79,
        66,
        65,
        82,
    ],
}
```
The above signature is "**FOOBAR**"

**Comment by wszdexdrf** [comment]
I personally think we can condense everything into one function and one macro invocation.
`pub struct HWISignature {
    #[serde(deserialize_with = "from_hex")]
    pub signature: Vec<u8>,
}

fn from_hex<'de, D>(deserializer: D) -> Result<Vec<u8>, D::Error>
where
    D: Deserializer<'de>,
{
    let s: String = Deserialize::deserialize(deserializer)?;
    Ok((0..s.len())
        .step_by(2)
        .map(|i| u8::from_str_radix(&s[i..i + 2], 16).unwrap())
        .collect())
}`

Minimal changes to the codebase.

**Comment by wszdexdrf** [comment]
HWI doesn't return a hex string, it returns a base64 string. Crate `base64` is already included in the library as a dependency, so only a single line is required to convert to Vec<u8>. `Base64::decode(s)`  😆

**Comment by wszdexdrf** [comment]
See comments [here](https://github.com/bitcoin-core/HWI/blob/master/hwilib/commands.py#L228). Also see [one of the implementations](https://github.com/bitcoin-core/HWI/blob/3fe369d0379212fae1c72729a179d133b0adc872/hwilib/devices/digitalbitbox.py#L567) of `HardwareWalletClient `using `unhexlify `before base64 encoding

**Comment by danielabrozzoni** [comment]
Closed by 76410513b4df818ccf96bc88dbc608e5b5031df3 🎉
